[FLINK-19114][python] Introduce expression DSL for Python Table API

dianfu opened the pull request:
## What is the purpose of the change

*This pull request introduces expression DSL for Python Table API*

## Brief change log

  - *Introduce class Expression class which defines functions on an expression*
  - *Introduce module expressions which contains a series of utility functions to create an expression*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added test_expression.py*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (No, I'll add doc in another PR as this PR is already very huge)

HuangXingBo commented:
@dianfu Thanks a lot  for the PR.  It looks good overall. I have one question whether it is necessary for Pyflink to support all methods to align Java methods. For example, Python provides a way of operator overloading, I think overloading `==` is enough, we don't need to support is_equal. What do you think?

dianfu commented:
@HuangXingBo Thanks for the comments. I think you raised a very good point. Removing the methods which could be covered by existing built-in Python methods seems fair enough for me as the built-in methods are more Pythonic. I have updated the PR and removed these kinds of methods.

WeiZhong94 reviewed (commented):
@dianfu Thanks for your PR! Looks good overall and I have left some comments which can further improve the user experience.

dianfu commented on flink-python/pyflink/table/expressions.py:
```diff
@@ -0,0 +1,535 @@
+    return _unary_op("$", name)
+
+
+def lit(v, data_type: DataType = None):
```
Good suggestion. I have created a separate JIRA to track this: https://issues.apache.org/jira/browse/FLINK-19116

dianfu commented:
@WeiZhong94 Thanks a lot for the review. Updated the PR. Could you take a further look at of the PR?

WeiZhong94 commented on flink-python/pyflink/table/expressions.py:
```diff
@@ -0,0 +1,543 @@
+
+def concat_ws(separator: 'Union[str, Expression[str]]',
+              first: 'Union[str, Expression[str]]',
+              *others: 'Union[str, Expression[str]]'):
```
This function seems always returns a 'Expression[str]'

WeiZhong94 commented on flink-python/pyflink/table/expressions.py:
```diff
@@ -0,0 +1,543 @@
+                                 [_get_java_expression(other) for other in others]))
+
+
+def uuid() -> Union[str, 'Expression[str]']:
```
This function seems always returns a 'Expression[str]'

WeiZhong94 commented on flink-python/pyflink/table/expressions.py:
```diff
@@ -0,0 +1,543 @@
+            add_version_doc(o[1], "1.12.0")
+
+
+def col(name: str):
```
We should add the return type hint for all the interfaces to fully utilize the IDE’s type checking.

WeiZhong94 commented on flink-python/pyflink/table/expressions.py:
```diff
@@ -0,0 +1,543 @@
+    return _binary_op("range", start, end)
+
+
+def and_(predicate0: Union[bool, 'Expression[bool]'],
```
Union[bool, 'Expression[bool]'] -> 'Union[bool, Expression[bool]]'?
It would be better if we unify the writing.

WeiZhong94 reviewed (commented):
@dianfu It looks much better now. Just some minor comments about the type hints.
